From 47fcac3b105b60eb1c770131df4967b0b230a23a Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Mon, 30 Oct 2023 19:25:27 -0400
Subject: [PATCH 41/55] Add a prop to use linear backlight control

---
 .../main/domain/interactor/ScreenBrightnessInteractorImpl.kt  | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/packages/SystemUI/pods/brightness/src/main/domain/interactor/ScreenBrightnessInteractorImpl.kt b/packages/SystemUI/pods/brightness/src/main/domain/interactor/ScreenBrightnessInteractorImpl.kt
index 100630466..118d6a1a3 100644
--- a/packages/SystemUI/pods/brightness/src/main/domain/interactor/ScreenBrightnessInteractorImpl.kt
+++ b/packages/SystemUI/pods/brightness/src/main/domain/interactor/ScreenBrightnessInteractorImpl.kt
@@ -16,6 +16,7 @@
 
 package com.android.systemui.brightness.domain.interactor
 
+import android.os.SystemProperties
 import com.android.settingslib.display.BrightnessUtils
 import com.android.systemui.brightness.data.model.LinearBrightness
 import com.android.systemui.brightness.data.repository.ScreenBrightnessRepository
@@ -86,6 +87,9 @@ constructor(
     }
 
     private suspend fun GammaBrightness.toLinearBrightness(): LinearBrightness {
+        if (SystemProperties.getBoolean("persist.sys.phh.linear_brightness", false)) {
+            return LinearBrightness(value / 65536.0f)
+        }
         val bounds = screenBrightnessRepository.getMinMaxLinearBrightness()
         return LinearBrightness(
             BrightnessUtils.convertGammaToLinearFloat(
-- 
2.54.0

