From 05a5a585bcac9ab40729372515ca106c8d58aa61 Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Tue, 7 May 2024 22:44:25 +0000
Subject: [PATCH] Dont abort if the DnsHelper failed to init on BPF-less kernel

---
 DnsProxyListener.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/DnsProxyListener.cpp b/DnsProxyListener.cpp
index f55d2c5..023e92c 100644
--- a/DnsProxyListener.cpp
+++ b/DnsProxyListener.cpp
@@ -717,7 +717,7 @@ DnsProxyListener::DnsProxyListener() : FrameworkListener(SOCKET_NAME) {
     mGetDnsNetIdCommand = std::make_unique<GetDnsNetIdCommand>();
     registerCmd(mGetDnsNetIdCommand.get());
 
-    if (ADnsHelper_init()) abort();
+    if (ADnsHelper_init()) LOG(ERROR) << "ADnsHelper_init failed, continuing without it";
 }
 
 void DnsProxyListener::Handler::spawn() {
-- 
2.54.0

