From 2517e1218d68a82e2eebe0984827f0b4dc6cd14e Mon Sep 17 00:00:00 2001
From: Ahmad Ansori Palembani <46041660+null2264@users.noreply.github.com>
Date: Sun, 2 Aug 2026 12:17:19 +0700
Subject: [PATCH 1/4] style: Move brightness slider to the bottom

---
 .../qs/composefragment/QSFragmentCompose.kt   | 80 +++++++++++--------
 1 file changed, 45 insertions(+), 35 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
index b32d7ad9654f1..0d6514dbfaf26 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
@@ -99,6 +99,7 @@ import com.android.compose.animation.scene.ElementMatcher
 import com.android.compose.animation.scene.MutableSceneTransitionLayoutState
 import com.android.compose.animation.scene.SceneKey
 import com.android.compose.animation.scene.SceneTransitionLayout
+import com.android.compose.animation.scene.SceneTransitionLayoutState
 import com.android.compose.animation.scene.content.state.TransitionState
 import com.android.compose.animation.scene.rememberMutableSceneTransitionLayoutState
 import com.android.compose.animation.scene.transitions
@@ -699,6 +700,42 @@ constructor(
         }
     }
 
+    @Composable
+    private fun BrightnessSlider(
+        viewModel: QSFragmentComposeViewModel,
+        layoutState: SceneTransitionLayoutState,
+    ) {
+        Box(
+            Modifier.systemGestureExclusionInShade(
+                enabled = {
+                    /*
+                     * While we are transitioning into QS (either from QQS
+                     * or from gone), the global position of the brightness
+                     * slider will change in every frame. This causes
+                     * the modifier to send a new gesture exclusion
+                     * rectangle on every frame. Instead, only apply the
+                     * modifier when this is settled.
+                     */
+                    layoutState.transitionState is TransitionState.Idle &&
+                        viewModel.isNotTransitioning
+                }
+            )
+        ) {
+            AlwaysDarkMode {
+                BrightnessSliderContainer(
+                    viewModel =
+                        containerViewModel.brightnessSliderViewModel,
+                    containerColors =
+                        ContainerColors(
+                            Color.Transparent,
+                            ContainerColors.defaultContainerColor,
+                        ),
+                    modifier = Modifier.fillMaxWidth(),
+                )
+            }
+        }
+    }
+
     @Composable
     private fun ContentScope.QuickQuickSettingsElement(modifier: Modifier = Modifier) {
         val qqsPadding = viewModel.qqsHeaderHeight
@@ -795,6 +832,7 @@ constructor(
                                 .padding(horizontal = qsHorizontalMargin())
                     ) {
                         QuickQuickSettingsLayout(
+                            brightness = { BrightnessSlider(viewModel, layoutState) },
                             tiles = Tiles,
                             media = Media,
                             mediaInRow = viewModel.qqsMediaInRow,
@@ -859,38 +897,7 @@ constructor(
                         Spacer(
                             modifier = Modifier.height { qqsPadding + qsExtraPadding.roundToPx() }
                         )
-                        val BrightnessSlider =
-                            @Composable {
-                                Box(
-                                    Modifier.systemGestureExclusionInShade(
-                                        enabled = {
-                                            /*
-                                             * While we are transitioning into QS (either from QQS
-                                             * or from gone), the global position of the brightness
-                                             * slider will change in every frame. This causes
-                                             * the modifier to send a new gesture exclusion
-                                             * rectangle on every frame. Instead, only apply the
-                                             * modifier when this is settled.
-                                             */
-                                            layoutState.transitionState is TransitionState.Idle &&
-                                                viewModel.isNotTransitioning
-                                        }
-                                    )
-                                ) {
-                                    AlwaysDarkMode {
-                                        BrightnessSliderContainer(
-                                            viewModel =
-                                                containerViewModel.brightnessSliderViewModel,
-                                            containerColors =
-                                                ContainerColors(
-                                                    Color.Transparent,
-                                                    ContainerColors.defaultContainerColor,
-                                                ),
-                                            modifier = Modifier.fillMaxWidth(),
-                                        )
-                                    }
-                                }
-                            }
+
                         val TileGrid =
                             @Composable {
                                 Box {
@@ -947,7 +954,7 @@ constructor(
                             QuickSettingsLayout(
                                 brightness =
                                     if (viewModel.isBrightnessSliderVisible) {
-                                        { BrightnessSlider() }
+                                        { BrightnessSlider(viewModel, layoutState) }
                                     } else {
                                         {}
                                     },
@@ -1460,6 +1467,7 @@ private fun ContentScope.MediaObject(
 @Composable
 @VisibleForTesting
 fun QuickQuickSettingsLayout(
+    brightness: @Composable () -> Unit,
     tiles: @Composable () -> Unit,
     media: @Composable () -> Unit,
     mediaInRow: Boolean,
@@ -1471,11 +1479,13 @@ fun QuickQuickSettingsLayout(
         ) {
             Box(modifier = Modifier.weight(1f)) { tiles() }
             Box(modifier = Modifier.weight(1f)) { media() }
+            Box(modifier = Modifier.weight(1f)) { brightness() }
         }
     } else {
         Column(verticalArrangement = spacedBy(dimensionResource(R.dimen.qs_tile_margin_vertical))) {
             tiles()
             media()
+            brightness()
         }
     }
 }
@@ -1494,7 +1504,6 @@ fun QuickSettingsLayout(
             verticalArrangement = spacedBy(QuickSettingsShade.Dimensions.Padding),
             horizontalAlignment = Alignment.CenterHorizontally,
         ) {
-            brightness()
             Row(
                 horizontalArrangement = spacedBy(QuickSettingsShade.Dimensions.Padding),
                 verticalAlignment = Alignment.CenterVertically,
@@ -1502,15 +1511,16 @@ fun QuickSettingsLayout(
                 Box(modifier = Modifier.weight(1f)) { tiles() }
                 Box(modifier = Modifier.weight(1f)) { media() }
             }
+            brightness()
         }
     } else {
         Column(
             verticalArrangement = spacedBy(QuickSettingsShade.Dimensions.Padding),
             horizontalAlignment = Alignment.CenterHorizontally,
         ) {
-            brightness()
             tiles()
             media()
+            brightness()
         }
     }
 }

From 30a4c0a79d541b2352e5bf998eba439308da5707 Mon Sep 17 00:00:00 2001
From: Ahmad Ansori Palembani <46041660+null2264@users.noreply.github.com>
Date: Sun, 2 Aug 2026 18:16:31 +0700
Subject: [PATCH 2/4] fix: containerViewModel -> viewModel.containerViewModel

---
 .../android/systemui/qs/composefragment/QSFragmentCompose.kt    | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
index 0d6514dbfaf26..b609d66bfa43d 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
@@ -724,7 +724,7 @@ constructor(
             AlwaysDarkMode {
                 BrightnessSliderContainer(
                     viewModel =
-                        containerViewModel.brightnessSliderViewModel,
+                        viewModel.containerViewModel.brightnessSliderViewModel,
                     containerColors =
                         ContainerColors(
                             Color.Transparent,

From ca63a11c46bbdb1d054eb10ac9cf0377dc675ac8 Mon Sep 17 00:00:00 2001
From: Ahmad Ansori Palembani <46041660+null2264@users.noreply.github.com>
Date: Sun, 2 Aug 2026 22:11:11 +0700
Subject: [PATCH 3/4] fix: Add missing transition

---
 .../qs/composefragment/QSFragmentCompose.kt       | 15 ++++++++++++---
 .../ui/FromQuickQuickSettingsToQuickSettings.kt   |  1 +
 .../systemui/qs/shared/ui/QuickSettings.kt        |  1 +
 3 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
index b609d66bfa43d..d400711c700dd 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
@@ -777,6 +777,11 @@ constructor(
                         }
                         .padding(top = { qqsPadding }, bottom = { bottomPadding })
             ) {
+                val BrightnessSlider: @Composable () -> Unit = {
+                    Element(Elements.BrightnessSlider, modifier = modifier) {
+                        BrightnessSlider(viewModel, layoutState)
+                    }
+                }
                 val Tiles =
                     @Composable {
                         // When always compose is false, this will always be true, and we'll be
@@ -832,7 +837,7 @@ constructor(
                                 .padding(horizontal = qsHorizontalMargin())
                     ) {
                         QuickQuickSettingsLayout(
-                            brightness = { BrightnessSlider(viewModel, layoutState) },
+                            brightness = BrightnessSlider,
                             tiles = Tiles,
                             media = Media,
                             mediaInRow = viewModel.qqsMediaInRow,
@@ -897,7 +902,11 @@ constructor(
                         Spacer(
                             modifier = Modifier.height { qqsPadding + qsExtraPadding.roundToPx() }
                         )
-
+                        val BrightnessSlider: @Composable () -> Unit = {
+                            Element(Elements.BrightnessSlider, modifier = modifier) {
+                                BrightnessSlider(viewModel, layoutState)
+                            }
+                        }
                         val TileGrid =
                             @Composable {
                                 Box {
@@ -954,7 +963,7 @@ constructor(
                             QuickSettingsLayout(
                                 brightness =
                                     if (viewModel.isBrightnessSliderVisible) {
-                                        { BrightnessSlider(viewModel, layoutState) }
+                                        { BrightnessSlider() }
                                     } else {
                                         {}
                                     },
diff --git a/packages/SystemUI/src/com/android/systemui/qs/composefragment/ui/FromQuickQuickSettingsToQuickSettings.kt b/packages/SystemUI/src/com/android/systemui/qs/composefragment/ui/FromQuickQuickSettingsToQuickSettings.kt
index 64cd529b7247c..ba7aa0e8a99af 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/composefragment/ui/FromQuickQuickSettingsToQuickSettings.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/composefragment/ui/FromQuickQuickSettingsToQuickSettings.kt
@@ -31,6 +31,7 @@ fun TransitionBuilder.quickQuickSettingsToQuickSettings(
     anchoredTranslate(Elements.QuickSettingsContent, Elements.GridAnchor)
 
     sharedElement(Elements.TileElementMatcher, enabled = animateTilesExpansion())
+    sharedElement(Elements.BrightnessSlider)
 
     // This will animate between 0f (QQS) and 0.5, fading in the QQS tiles when coming back
     // from non first page QS. The QS content ends fading out at 0.43f, so there's a brief
diff --git a/packages/SystemUI/src/com/android/systemui/qs/shared/ui/QuickSettings.kt b/packages/SystemUI/src/com/android/systemui/qs/shared/ui/QuickSettings.kt
index ca3626594f4bb..8ebf07a11fe62 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/shared/ui/QuickSettings.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/shared/ui/QuickSettings.kt
@@ -33,6 +33,7 @@ object QuickSettings {
 
         val QuickQuickSettingsAndMedia = ElementKey("QuickQuickSettingsAndMedia")
         val SplitShadeQuickSettings = ElementKey("SplitShadeQuickSettings")
+        val BrightnessSlider = ElementKey("BrightnessSlider")
     }
 
     object SharedValues {

From 3cf395cbe8d60fc18bf3395a7ac18690f22d2ccf Mon Sep 17 00:00:00 2001
From: Ahmad Ansori Palembani <46041660+null2264@users.noreply.github.com>
Date: Mon, 3 Aug 2026 20:00:12 +0700
Subject: [PATCH 4/4] style: Media should be shown below brightness slider
 unless on landscape mode

---
 .../qs/composefragment/QSFragmentCompose.kt   | 20 ++++++++++---------
 1 file changed, 11 insertions(+), 9 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
index d400711c700dd..be8b31f30950d 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/composefragment/QSFragmentCompose.kt
@@ -1482,19 +1482,21 @@ fun QuickQuickSettingsLayout(
     mediaInRow: Boolean,
 ) {
     if (mediaInRow) {
-        Row(
-            horizontalArrangement = spacedBy(dimensionResource(R.dimen.qs_tile_margin_vertical)),
-            verticalAlignment = Alignment.CenterVertically,
-        ) {
-            Box(modifier = Modifier.weight(1f)) { tiles() }
-            Box(modifier = Modifier.weight(1f)) { media() }
-            Box(modifier = Modifier.weight(1f)) { brightness() }
+        Column(verticalArrangement = spacedBy(dimensionResource(R.dimen.qs_tile_margin_vertical))) {
+            Row(
+                horizontalArrangement = spacedBy(dimensionResource(R.dimen.qs_tile_margin_vertical)),
+                verticalAlignment = Alignment.CenterVertically,
+            ) {
+                Box(modifier = Modifier.weight(1f)) { tiles() }
+                Box(modifier = Modifier.weight(1f)) { media() }
+            }
+            brightness()
         }
     } else {
         Column(verticalArrangement = spacedBy(dimensionResource(R.dimen.qs_tile_margin_vertical))) {
             tiles()
-            media()
             brightness()
+            media()
         }
     }
 }
@@ -1528,8 +1530,8 @@ fun QuickSettingsLayout(
             horizontalAlignment = Alignment.CenterHorizontally,
         ) {
             tiles()
-            media()
             brightness()
+            media()
         }
     }
 }
