From b2b61de50812698a93009fa6913bb1df4fbb7493 Mon Sep 17 00:00:00 2001
From: Andy CrossGate Yan <GeForce8800Ultra@gmail.com>
Date: Sun, 26 May 2024 21:34:17 +0800
Subject: [PATCH 13/16] SurfaceFlinger: Disable SF HWC backpressure

---
 services/surfaceflinger/SurfaceFlinger.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index c957090..3972e28 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -3056,7 +3056,8 @@ bool SurfaceFlinger::commit(PhysicalDisplayId pacesetterId,
         }
     }
 
-    if (pacesetterFrameTargetPtr->wouldBackpressureHwc()) {
+    if (base::GetBoolProperty("persist.sys.phh.enable_sf_hwc_backpressure"s, true)
+            && pacesetterFrameTargetPtr->isFramePending()) {
         if (mBackpressureGpuComposition || pacesetterFrameTargetPtr->didMissHwcFrame()) {
             mScheduler->getVsyncSchedule()->getTracker().onFrameMissed(
                     pacesetterFrameTargetPtr->expectedPresentTime());
-- 
2.54.0

